File: script/generate_data.py
import os
import json
from bs4 import BeautifulSoup
from datetime import datetime

# 读取 bookmarks.html 文件
try:
    with open('bookmarks.html', 'r', encoding='utf-8') as f:
        content = f.read()
except Exception as e:
    print(f"Failed to read file: {e}")
    exit()

# 去除所有的 <p> 和 <DT> 标签
content = content.replace('<p>', '').replace('<DT>', '')

# 使用 BeautifulSoup 解析 HTML 内容
soup = BeautifulSoup(content, 'html.parser')

# 递归函数来提取书签数据
def extract_bookmarks(dl):
    bookmarks = []
    for element in dl.children:
        if element.name == 'h3':  # 文件夹名称
            folder_name = element.get_text(strip=True)
            folder_content = []

            # 查找该文件夹下的 dl 标签并递归提取
            next_dl = element.find_next('dl')
            if next_dl:
                folder_content.extend(extract_bookmarks(next_dl))

            bookmarks.append({'fname': folder_name, 'content': folder_content})
        elif element.name == 'a':  # 链接
            link_name = element.get_text(strip=True)
            link_url = element.get('href', '')
            bookmarks.append({'name': link_name, 'url': link_url})

    return bookmarks

# 查找根部的 dl 标签并开始解析
root_dl = soup.find_all('dl')[1] #选择第二个dl标签作为根
bookmarks = extract_bookmarks(root_dl)

# 获取当前时间并格式化
current_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')

# 输出为 JSON 格式
json_output = json.dumps(bookmarks, indent=4, ensure_ascii=False)

# 保存到文件
try:
    with open('data.json', 'w', encoding='utf-8') as f:
        f.write(json_output)
    print(f"[{current_time}] The data has been successfully saved as data.json")
except Exception as e:
    print(f"Failed to write file: {e}")

File: src/generate_data.js
const fs = require('fs');
const cheerio = require('cheerio');

// 读取 bookmarks.html 文件
fs.readFile('bookmarks.html', 'utf8', (err, content) => {
  if (err) {
    console.error('Failed to read file:', err);
    return;
  }

  // 去除所有的 <p> 和 <DT> , 防止干扰
  content = content.replace(/<\/?(p|DT)[^>]*>/g, '');

  // 使用 cheerio 解析修改后的 HTML 内容
  const $ = cheerio.load(content);

  // 递归函数来提取书签数据
  function extractBookmarks(dl) {
    const bookmarks = [];
    $(dl).children().each((index, element) => {
      if ($(element).is('h3')) {  // 文件夹名称
        const folderName = $(element).text().trim();
        const folderContent = [];

        // 查找该文件夹下的 dl 标签并递归提取
        const nextDl = $(element).next('dl');
        if (nextDl.length > 0) {
          folderContent.push(...extractBookmarks(nextDl));
        }

        bookmarks.push({ fname: folderName, content: folderContent });
      } else if ($(element).is('a')) {  // 链接
        const linkName = $(element).text().trim();
        const linkUrl = $(element).attr('href') || '';
        bookmarks.push({ name: linkName, url: linkUrl });
      }
    });
    return bookmarks;
  }

  // 查找 dl 标签并开始解析
  const rootDl = $('dl').eq(1);  // 选择第二个 dl 标签作为根部
  const bookmarks = extractBookmarks(rootDl);

  // 获取当前时间并格式化
  const current_time = new Date().toISOString().replace('T', ' ').slice(0, 19);

  // 输出为 JSON 格式
  const jsonOutput = JSON.stringify(bookmarks, null, 4);

  // 保存到文件
  fs.writeFileSync('data.json', jsonOutput, 'utf-8');
  console.log(`[${current_time}]The data has been successfully saved as data.json`);
});